### PR TITLE
Issue #11579 - Reject Target in Request Path

### LINKED ISSUE #11579: Introduce `UriCompliance.Violation.FRAGMENT` to reject HTTP Request Line that includes fragment section.
**Jetty version(s)**
12.0.7

**Jetty Environment**
All

**Java version/vendor** `(use: java -version)`
All

**OS type/version**
All

**Description**
While working PR #11496 the idea of not allowing FRAGMENT section in a Request Line was introduced.

It is good idea that seems to follow the HTTP spec.

* https://datatracker.ietf.org/doc/html/rfc9110#section-4.1 - indicates that the relative URI shouldn't have the fragment component
* https://datatracker.ietf.org/doc/html/rfc9110#section-4.2.5 - says that the fragment component is not part of the ABNF for http or https URIs
* https://datatracker.ietf.org/doc/html/rfc9110#section-7.1 - says that the server should ignore/drop the fragment component, as the fragment component is reserved for client-side processing
* https://datatracker.ietf.org/doc/html/rfc9110#section-10.2.2 - points out that a redirect `Location` response header can contain a fragment indicator
* https://datatracker.ietf.org/doc/html/rfc9110#section-17.11 - there is a security concern as well with exposure of fragment after redirect

If we do this, we should be careful how we do it, and allow a configurable UriCompliance mode to configure the behavior.

See original commit (reverted in PR #11496):
 https://github.com/jetty/jetty.project/pull/11496/commits/fed10f79aacecfcb4118bcbe39fcb21d98666f25

### PR DESCRIPTION
**Pull request by joakime**
Initial implementation of rejecting a URI Target in the Request Path.

### CODE REVIEW

**Comment by gregw** [comment]
@joakime status?

**Comment by joakime** [comment]
I haven't look at this in a while, since we moved it here from PR #11496
I'll update this PR this week.

**Review by gregw — CHANGES_REQUESTED** [review]
This is a placeholder review waiting for progress on this.  Please request re-review when it is ready.

**Comment by joakime** [comment]
rebase against 12.1.x

File: jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java [hunk 1]
```diff
@@ -1452,6 +1457,9 @@ else if (c == '/')
                 }
             }
 
+            if (last != null && state.ordinal() > last.ordinal())
+                throw new IllegalArgumentException("uri cannot go beyond " + last);
+
```
Comment (gregw) [inline jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:1462]: I don't think we should not parse past the fragment etc.  We should parse to the end, but record a violation if there is a fragment.

**Review by gregw — CHANGES_REQUESTED** [review]
Let me have a go at this from the compliance end of things...

**Comment by joakime** [comment]
Closing in favor of PR #12504

**Pull request closed without merging**